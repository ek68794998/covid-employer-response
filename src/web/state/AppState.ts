import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}