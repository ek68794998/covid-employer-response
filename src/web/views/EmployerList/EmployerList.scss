@import "../Colors.scss";
@import "../ReactModal.scss";

$card-min-width: 374px;
$close-button-size: 48px;
$close-button-size-small: 32px;

$body-2column-width: ($card-min-width * 2 + 40px);
$body-3column-width: ($card-min-width * 3 + 120px);

div.EmployerList__Container--Loading, div.EmployerList__Container--NoResults {
	font: 400 20px 'Baloo Chettan 2';
	margin: 16px auto;
	text-align: center;
}

div.EmployerList__Container {
	display: grid;
	grid-auto-rows: max-content;
	grid-gap: 16px;
	grid-template-columns: repeat(3, 1fr);
}

div.EmployerList__Item {
	height: 100%;
}

button.EmployerList__CloseModal {
	background: $background;
	border: 0;
	border-radius: ($close-button-size / 2);
	box-shadow: 0 2px 6px $box-shadow-color;
	cursor: pointer;
	position: fixed;
	height: $close-button-size;
	outline: none;
	pointer-events: initial;
	right: ($page-padding / 2);
	top: ($page-padding / 2);
	width: $close-button-size;
	z-index: 21;

	i {
		font-size: 36px;
		margin-top: 2px;
	}
}

@media (max-width: $body-2column-width) {
	div.EmployerList__Container {
		grid-template-columns: repeat(1, 1fr);
	}

	button.EmployerList__CloseModal {
		height: $close-button-size-small;
		right: ($page-padding-narrow / 2);
		top: ($page-padding-narrow / 2);
		width: $close-button-size-small;

		i {
			font-size: 24px;
			margin-left: -2px;
		}
	}
}

@media (max-width: $body-3column-width) and (min-width: $body-2column-width) {
	div.EmployerList__Container {
		grid-template-columns: repeat(2, 1fr);
	}
}
