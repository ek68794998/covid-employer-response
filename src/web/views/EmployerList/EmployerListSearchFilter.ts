import { EmployerRecord } from "../../../common/EmployerRecord";

export class EmployerListSearchFilter {
	public text: string = "";

	public static isMatch(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		return e.name.toLowerCase().indexOf(f.text.toLowerCase()) >= 0;
	}
}