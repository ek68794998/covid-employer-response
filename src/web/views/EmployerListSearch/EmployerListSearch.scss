@import "../Colors.scss";
@import "../Dimensions.scss";

$input-buffer: 8px;
$input-height: 40px;

.EmployerListSearch__Container {
	align-items: center;
	display: flex;
}

.EmployerListSearch__SearchContainer {
	align-items: center;
	border: 1px solid $outline-color;
	border-radius: ($input-height / 2);
	display: flex;
	flex: 1;
	flex-direction: column;
	font: 400 16px 'Baloo Chettan 2';

	input, button {
		font: inherit;
	}
}

.EmployerListSearch__ViewTypeContainer {
	margin-left: 8px;

	button {
		background: none;
		border: none;
		border-radius: ($input-height / 2);
		cursor: pointer;
		height: $input-height;
		margin-left: 4px;
		outline: none;
		padding: 8px;
		transition: 0.1s all;
		width: $input-height;

		&:hover {
			background: rgba(0, 0, 0, 0.1);
		}
	}
}

.EmployerListSearch__InputContainer {
	display: flex;
	flex: 1;
	height: 100%;
	width: 100%;

	> i {
		opacity: 0.5;
	}

	input[type=search] {
		background: none;
		border: 0;
		font: 400 16px 'Baloo Chettan 2';
		outline: 0;
		width: 100%;
	}

	.EmployerListSearch__Input {
		align-items: center;
		border-right: 1px solid $outline-color;
		display: flex;
		flex: 1;
		padding: $input-buffer;
		white-space: nowrap;

		input {
			margin-left: 4px;
		}
	}

	.EmployerListSearch__ExpandFilters {
		align-items: center;
		background: none;
		border: none;
		cursor: pointer;
		display: flex;
		line-height: 0;
		margin: 0;
		outline: none;
		padding-right: 16px;

		&.EmployerListSearch__ExpandFilters--Open {
			color: $link-normal;
		}

		i {
			margin: 0 8px 0 4px;
		}
	}
}

.EmployerListSearch__FilterContainer {
	align-items: center;
	border-top: 1px solid $outline-color;
	display: flex;
	flex: 0;
	height: 100%;
	padding: $input-buffer;
	width: 100%;

	> div {
		margin-right: 8px;
	}
}
