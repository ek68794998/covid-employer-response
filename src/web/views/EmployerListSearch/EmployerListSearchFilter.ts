import { EmployerRecord } from "../../../common/EmployerRecord";

export class EmployerListSearchFilter {
	public static readonly SMALL_BOUNDARY: number = 1000;

	public static readonly LARGE_BOUNDARY: number = 100000;

	public international: boolean = true;

	public large: boolean = true;

	public medium: boolean = true;

	public national: boolean = true;

	public small: boolean = true;

	public text: string = "";

	public static isMatch(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!this.isMatchForLocation(f, e)) {
			return false;
		}

		if (!this.isMatchForEmployeeCount(f, e)) {
			return false;
		}

		return this.isMatchForFullText(f, e);
	}

	private static isMatchForEmployeeCount(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!e.employeesBefore) {
			return false;
		}

		const getNumberIsBetween = (v: number, lower: number, upper: number): boolean =>
			v >= lower && v <= upper;

		const getNumberIsMatch = (v: number): boolean => {
			if (!f.small && v <= this.SMALL_BOUNDARY) {
				return false;
			}

			if (!f.large && v >= this.LARGE_BOUNDARY) {
				return false;
			}

			if (!f.medium
				&& getNumberIsBetween(v, this.SMALL_BOUNDARY, this.LARGE_BOUNDARY)) {

				return false;
			}

			return true;
		};

		const lowerBound: number = e.employeesBefore.lowerBound || 0;
		const upperBound: number = Math.max(e.employeesBefore.upperBound, lowerBound);

		if (e.employeesBefore.type === "approximately"
			|| e.employeesBefore.type === "exactly"
			|| lowerBound <= 0) {

			return getNumberIsMatch(upperBound);
		}

		if (upperBound <= 0) {
			return getNumberIsMatch(lowerBound);
		}

		const rangeIncludesSmall: boolean = lowerBound < this.SMALL_BOUNDARY;

		const rangeIncludesMedium: boolean =
			getNumberIsBetween(lowerBound, this.SMALL_BOUNDARY, this.LARGE_BOUNDARY)
			|| getNumberIsBetween(upperBound, this.SMALL_BOUNDARY, this.LARGE_BOUNDARY);

		const rangeIncludesLarge: boolean = upperBound > this.LARGE_BOUNDARY;

		return (f.small && rangeIncludesSmall)
			|| (f.medium && rangeIncludesMedium)
			|| (f.large && rangeIncludesLarge);
	}

	private static isMatchForFullText(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!f.text) {
			return true;
		}

		const fieldsToSearch: (string | undefined)[] = [
			e.name.toLowerCase(),
			e.location?.city.toLowerCase(),
			e.shortName?.toLowerCase(),
			e.ticker?.toLowerCase(),
		];

		if (e.aliases) {
			e.aliases.forEach((a: string) => fieldsToSearch.push(a));
		}

		return fieldsToSearch.some((field?: string) => field && field.indexOf(f.text.toLowerCase()) >= 0);
	}

	private static isMatchForLocation(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!e.location) {
			return false;
		}

		if ((!f.international && e.location.international) || (!f.national && !e.location.international)) {
			return false;
		}

		return true;
	}
}