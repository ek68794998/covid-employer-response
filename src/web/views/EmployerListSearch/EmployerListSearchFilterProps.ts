import { RouteProps } from "react-router-dom";
import { EmployerListSearchFilter } from "./EmployerListSearchFilter";

export interface EmployerListSearchFilterProps extends RouteProps {
	initialFilter: EmployerListSearchFilter;

	onUpdateFilterValue: (filters: Partial<EmployerListSearchFilter>) => void;
}
