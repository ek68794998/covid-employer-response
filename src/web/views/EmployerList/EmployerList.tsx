import React, { useState } from "react";
import { useSelector } from "react-redux";
import { RouteProps } from "react-router-dom";

import { EmployerRecordMetadata } from "../../../common/EmployerRecordMetadata";
import { format, LocalizedStrings } from "../../../common/LocalizedStrings";
import { getEmployersList } from "../../state/ducks/employers/selectors";
import { getStrings } from "../../state/ducks/localization/selectors";

import EmployerListItem from "../EmployerListItem/EmployerListItem";
import { EmployerListSearchFilter } from "../EmployerListSearch/EmployerListSearchFilter";

import "./EmployerList.scss";

const loadChunkSize: number = 3;
const loadMoreRowCount: number = 4;
const loadMoreCount: number = loadChunkSize * loadMoreRowCount;

interface Props extends RouteProps {
	searchFilter: EmployerListSearchFilter;
}

const EmployerList: React.FC<Props> = (props: Props): React.ReactElement => {
	const strings: LocalizedStrings = useSelector(getStrings);
	const employersList: EmployerRecordMetadata[] | undefined = useSelector(getEmployersList);

	const [ visibleItemCount, setVisibleItemCount ] = useState(loadMoreCount);

	const { searchFilter } = props;

	if (!employersList) {
		return (
			<div className="EmployerList__Container--Loading">
				{strings.loading}
			</div>
		);
	}

	const filteredEmployers: EmployerRecordMetadata[] =
		employersList
			.filter((e: EmployerRecordMetadata) => EmployerListSearchFilter.isMatch(searchFilter, e))
			.sort((a: EmployerRecordMetadata, b: EmployerRecordMetadata) => {
				const nameMatcher: RegExp = /^(The |A )?(.*)$/i;

				const aName: string | undefined = nameMatcher.exec(a.shortName || a.name)?.[2];
				const bName: string | undefined = nameMatcher.exec(b.shortName || b.name)?.[2];

				return (aName || "").localeCompare(bName || "");
			});

	if (!filteredEmployers.length) {
		return (
			<div className="EmployerList__Container--NoResults">
				{strings.noResults}
			</div>
		);
	}

	return (
		<div className="EmployerList__Container">
			<div className="EmployerList__Grid">
				{filteredEmployers.slice(0, visibleItemCount).map(
					(e: EmployerRecordMetadata, i: number): JSX.Element => (
						<div className="EmployerList__Item" key={`${i}-${e.id}`}>
							<EmployerListItem employerId={e.id} showDetails={true} />
						</div>
					))}
			</div>
			{visibleItemCount < filteredEmployers.length && (
				<button className="EmployerList__LoadMore" onClick={(): void => setVisibleItemCount(visibleItemCount + loadMoreCount)}>
					{format(
						strings.loadMore,
						{
							count: Math.min(filteredEmployers.length - visibleItemCount, loadMoreCount),
						})}
				</button>
			)}
		</div>
	);
};

export default EmployerList;
