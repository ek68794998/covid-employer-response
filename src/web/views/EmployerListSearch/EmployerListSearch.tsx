import React, { useEffect, useState, useContext } from "react";
import { useSelector } from "react-redux";

import { DesignHelpers } from "../../../common/DesignHelpers";
import { LocalizedStrings } from "../../../common/LocalizedStrings";

import { getStrings } from "../../state/ducks/localization/selectors";

import { EmployerRouteContext, EmployerRouteContextData } from "../EmployerRoute/EmployerRouteContext";

import EmployeeCountFilterControl from "./EmployerListFilterControl/EmployeeCountFilterControl";
import InternationalTypeFilterControl from "./EmployerListFilterControl/InternationalTypeFilterControl";
import { EmployerListSearchFilter } from "./EmployerListSearchFilter";

import "./EmployerListSearch.scss";

const EmployerListSearch: React.FC = (): React.ReactElement => {
	const strings: LocalizedStrings = useSelector(getStrings);

	const listContext: EmployerRouteContextData = useContext(EmployerRouteContext);

	const [ listViewMode, setListViewMode ] = useState(listContext.listViewMode);
	const [ searchFilters, setSearchFilters ] = useState(listContext.searchFilters);
	const [ isFiltersetVisible, setIsFiltersetVisible ] = useState(false);

	useEffect(
		(): void => {
			listContext.setSearchFilters(searchFilters);
		},
		[ searchFilters ]);

	useEffect(
		(): void => {
			listContext.setListViewMode(listViewMode);
		},
		[ listViewMode ]);

	const updateSearchFilters = (updates?: Partial<EmployerListSearchFilter>): void => {
		setSearchFilters({ ...searchFilters, ...updates });
	};

	const onInput =
		(e: React.FormEvent<HTMLInputElement>): void => updateSearchFilters({ text: e.currentTarget.value });

	return (
		<div className="EmployerListSearch__Container">
			<div className="EmployerListSearch__SearchContainer">
				<div className="EmployerListSearch__InputContainer">
					<div className="EmployerListSearch__Input">
						{DesignHelpers.materialIcon("search")}
						<input
							defaultValue={searchFilters.text}
							onInput={onInput}
							placeholder={strings.search}
							type="search"
						/>
					</div>
					<button
						className={`EmployerListSearch__ExpandFilters EmployerListSearch__ExpandFilters--${isFiltersetVisible ? "Open" : "Closed"}`}
						onClick={(): void => setIsFiltersetVisible(!isFiltersetVisible)}
					>
						{DesignHelpers.materialIcon("filter_list")}
						{strings.filter}
					</button>
				</div>
				{isFiltersetVisible && (
					<div className="EmployerListSearch__FilterContainer">
						<InternationalTypeFilterControl filter={searchFilters} onUpdateFilter={updateSearchFilters} />
						<EmployeeCountFilterControl filter={searchFilters} onUpdateFilter={updateSearchFilters} />
					</div>
				)}
			</div>
			<div className="EmployerListSearch__ViewTypeContainer">
				<button
					onClick={(): void => setListViewMode("rows")}
				>
					{DesignHelpers.materialIcon("view_headline")}
				</button>
				<button
					onClick={(): void => setListViewMode("cards")}
				>
					{DesignHelpers.materialIcon("view_module")}
				</button>
			</div>
		</div>
	);
};

export default EmployerListSearch;
