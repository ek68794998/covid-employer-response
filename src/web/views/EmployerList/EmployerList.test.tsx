import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { mockComponent } from "../../../__tests__/TestUtils";
import { EmployerRecord } from "../../../common/EmployerRecord";

import { AppState } from "../../state/AppState";
import configureStore from "../../state/configureStore";

import { EmployerListSearchFilter } from "../EmployerListSearch/EmployerListSearchFilter";

import EmployerList from "./EmployerList";

jest.mock(
	"../EmployerListDetails/EmployerListDetails",
	() => mockComponent("EmployerListDetails"));

jest.mock(
	"../EmployerPageDetails/EmployerPageDetails",
	() => mockComponent("EmployerPageDetails"));

describe("<EmployerList />", () => {
	test("renders without exploding", () => {
		const store: Store<AppState, AnyAction> = configureStore({
			strings: {
				loading: "ℓôáδïñϱ",
				noResults: "ñôRèƨúℓƭƨ",
			},
		});

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployerList
							employers={[ new EmployerRecord() ]}
							searchFilter={new EmployerListSearchFilter()}
						/>
					</BrowserRouter>
				</Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});
});
