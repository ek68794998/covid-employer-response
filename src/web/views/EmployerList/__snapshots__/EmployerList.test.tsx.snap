// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmployerList /> renders empty list of employers 1`] = `
<div
  className="EmployerList__Container--NoResults"
>
  ñôRèƨúℓƭƨ
</div>
`;

exports[`<EmployerList /> renders list of employers 1`] = `
<div
  className="EmployerList__Container"
>
  <div
    className="EmployerList__Item"
  >
    [component: EmployerListDetails employer=([object Object]) onClick=([Function]) /]
  </div>
  <div
    className="EmployerList__Item"
  >
    [component: EmployerListDetails employer=([object Object]) onClick=([Function]) /]
  </div>
</div>
`;

exports[`<EmployerList /> renders without exploding 1`] = `
<div
  className="EmployerList__Container--Loading"
>
  ℓôáδïñϱ
</div>
`;
